src/ape_cache/query.py

```diff
@@ -0,0 +1,472 @@
+                if not result:
+                    return None
+
+                return self._compute_estimate(query, result)
+
+        except QueryEngineError as err:
+            logger.warning(f"Cannot perform query on cache database: {err}")
+            return None
+
+    @singledispatchmethod
+    def _perform_query_clause(self, query: QueryType) -> Select:
+        """
+        Create the SQLAlchemy select statement to perform the query.
+
+        Args:
+            query (QueryType): Choice of query type to perform a
+                check of the number of rows that match the clause.
+
+        Raises:
+            :class:`~ape.exceptions.QueryEngineError`: When given an
+                incompatible QueryType
+
+        Returns:
+            `sqlalchemy.sql.expression.TextClause`
+        """
+
+        raise QueryEngineError(
+            "Not a compatible QueryType. For more details see our docs "
+            "https://docs.apeworx.io/ape/stable/methoddocs/"
+            "exceptions.html#ape.exceptions.QueryEngineError"
+        )
+
+    @_perform_query_clause.register
+    def _perform_block_clause(self, query: BlockQuery) -> Select:
+        return (
+            select([column(c) for c in query.columns])
+            .where(Blocks.number >= query.start_block)
+            .where(Blocks.number <= query.stop_block)
+            .where(Blocks.number % query.step == 0)
+        )
+
+    @_perform_query_clause.register
+    def _perform_transaction_clause(self, query: BlockTransactionQuery) -> Select:
+        return select([Transactions]).where(Transactions.block_hash == query.block_id)
+
+    @_perform_query_clause.register
+    def _perform_contract_event_clause(self, query: ContractEventQuery) -> Select:
+        return (
+            select([column(c) for c in query.columns])
+            .where(ContractEvents.block_number >= query.start_block)
+            .where(ContractEvents.block_number <= query.stop_block)
+            .where(ContractEvents.block_number % query.step == 0)
+        )
+
+    @singledispatchmethod
+    def perform_query(self, query: QueryType) -> Optional[Iterator]:  # type: ignore
+        """
+        Performs the requested query from cache.
+
+        Args:
+            query (QueryType): Choice of query type to perform a
+                check of the number of rows that match the clause.
+
+        Raises:
+            :class:`~ape.exceptions.QueryEngineError`: When given an
+                incompatible QueryType
+
+        Returns:
+            Optional[Iterator]
+        """
+
+        raise QueryEngineError(
+            "Not a compatible QueryType. For more details see our docs "
+            "https://docs.apeworx.io/ape/stable/methoddocs/"
+            "exceptions.html#ape.exceptions.QueryEngineError"
+        )
+
+    @perform_query.register
+    def _perform_block_query(self, query: BlockQuery) -> Optional[Iterator[BlockAPI]]:
+        try:
```

Comment:
Remove the try-except and shift over, and then it no longer needs to be `Optional`
```suggestion
    def _perform_block_query(self, query: BlockQuery) -> Iterator[BlockAPI]:
```

Reply:
Note that under normal circumstances, there should never be an error here if the `estimate_query` logic is performing correctly, so it's actually a good thing that `QueryEngineError`s bubble up to the user, because that alerts us to some sort of issue with our logic

Reply:
[21465f7](https://github.com/ApeWorX/ape/pull/680/commits/21465f70f97a56ad08ed00e35500fc40864eef1d)